bluefin-internal/src/Bluefin/Internal.hs

```diff
@@ -388,6 +388,20 @@ catch ::
   Eff es a
 catch f h = handle h f
 
+-- | @'bracket' acquire release thing@ : @acquire@ a resource, do a @thing@
+-- with it, and @release@ the resource even if @thing@ threw an exception.
+--
+-- For more details, see 'Control.Exception.bracket'.
+bracket ::
+  io :> es =>
+  IOE io ->
```

Comment:
Can't we do without this? I know this operation is not technically pure, since there are various ways we could inspect the exception, but it seems morally pure to me. On the other hand, if we're allocating resources that need `bracket` to cleanly release then we were probably using `IO` anyway, so maybe it doesn't matter.

Reply:
Perhaps we can include this as `bracketIO`, because we're confident it's safe, and then consider whether `bracket`, without `IOE`, also warrants inclusion?

Reply:
How about an effect `MayFail :: Effects -> Type` which is a subeffect of `IOE` and `Exception`?

Reply:
Do you mean "subeffect" in the sense that there are injections `IOE e -> MayFail e` and `Exception t e -> MayFail e`?

Reply:
Indeed! See my edit.

Reply:
What would happen if we do something like

```.hs
try $ \ex -> bracket ex before after thing
```

where `before`, `after` and `thing` do not refer to `ex`? Would bracket do any cleanup at all on aborted `thing`?

Reply:
Maybe the right way of thinking about this is that morally, when we bring `IOE` or `Exception e` into scope, `Eff` is augmented with that knowledge and `bracket` can use it to be aware of all possible exception types that could be thrown in its `thing`. That seems pretty convincing that `bracket` should not be depending directly upon any effect. `Eff` (morally) already knows the relevant ones!

Reply:
My proposal is that  `MayFail` represents the knowledge that some exception may be thrown, without knowing what that exception is.

`try $ \ex -> bracket ex before after thing`, may still execute `after` because what `bracket` actually uses is not `ex` as a whole (which `thing` has no knowledge of) but a fragment of `ex` that's present in all `ex` handles (which `thing` may thus also possess via another handle).

Reply:
OK, but if it's in all `Exception` handles then isn't this an equivalent version of `bracket` but unconditional?

```.hs
bracket' :: Eff es a -> (a -> Eff es ()) -> (a -> Eff es b) -> Eff es b
bracket' before after thing =
  fmap (either id absurd) $ try $ \ex -> bracket ex before after thing
```

Reply:
That's a good one.